### PR TITLE
[pack] Key Vault secret repository

### PR DESCRIPTION
**Pull request by alrod**
* introduced base class for secret repository to handle sentinels
* ISecretsRepository is changed to use "ScriptSecrets" instead JSONstring
* Normalization example: '%'->"ASCII045"

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs [hunk 1]
```diff
@@ -55,6 +55,12 @@ internal ISecretsRepository CreateSecretsRepository()
             {
                 return new FileSystemSecretsRepository(_options.CurrentValue.SecretsPath);
             }
+            else if (secretStorageType != null && secretStorageType.Equals("Microsoft.KeyVault", StringComparison.OrdinalIgnoreCase))
```
[the remark this answers is not in the document]
Comment (alrod) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs]: @mattchenderson suggested this as in onenote

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+            }
+            await Task.WhenAll(setTasks);
+
+            string filePath = GetSecretsSentinelFilePath(type, functionName);
+            await FileUtility.WriteAsync(filePath, DateTime.UtcNow.ToString());
+        }
+
+        public async Task PurgeOldSecretsAsync(IList<string> currentFunctions, ILogger logger)
+        {
+            // no-op - allow stale secrets to remain
```
[the remark this answers is not in the document]
Comment (alrod) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: Same as BlobStoreageSecretsRepository. By the way it looks like we do not call `PurgeOldSecretsAsync` in V2.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+        private string GetVaultBaseUrl()
+        {
+            return $"https://{_vaultName}.vault.azure.net";
+        }
+
+        public static Dictionary<string, string> GetDictionaryFromScriptSecrets(ScriptSecrets secrets, string functionName)
+        {
+            Dictionary<string, string> dic = new Dictionary<string, string>();
+            if (secrets is HostSecrets)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: change to `secrets is HostSecrets hostSecrets` here and remove the cast below

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                {
+                    foreach (Key key in hostSecrets.SystemKeys)
+                    {
+                        dic.Add($"{systemKeyPrefix}{Normalize(key.Name)}", key.Value);
+                    }
+                }
+            }
+            else
+            {
+                FunctionSecrets functionSecrets = secrets as FunctionSecrets;
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: This implementation assumes `ScriptSecrets` will always be either `HostSecrets` or `FunctionSecrets`, checking that this cast succeeded here would help avoid issues if that changes.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                    result += normalizedCharMark + charValue;
+                }
+            }
+            return result;
+        }
+
+        public static string Denormalize(string normalizedName)
+        {
+            string result = string.Empty;
+            for (int index = 0; index < normalizedName.Length; index++)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: This logic can probably be simplified with a regex replace call that parses the match value into the character

File: src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultKeyValueConverterFactory.cs [hunk 6]
```diff
@@ -16,9 +16,9 @@ public sealed class DefaultKeyValueConverterFactory : IKeyValueConverterFactory
         private readonly bool _encryptionSupported;
         private static readonly PlaintextKeyValueConverter PlaintextValueConverter = new PlaintextKeyValueConverter(FileAccess.ReadWrite);
 
-        public DefaultKeyValueConverterFactory()
+        public DefaultKeyValueConverterFactory(bool isEncryptionSupported)
         {
-            _encryptionSupported = IsEncryptionSupported();
+            _encryptionSupported = !isEncryptionSupported && IsEncryptionSupported();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultKeyValueConverterFactory.cs]: nit: This flag name is a bit confusing. Perhaps change it to something like `allowEncryption` or something along those lines and add a documentation stating that setting that argument to true will cause the factory to use encryption if supported by the current environment.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs (hunk 5 above)
Comment (alrod) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: I tried but it looks to hard. I do not want to spend lot of time on it.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs]: I agree that we should just use "keyvault"

File: src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+        public override async Task PurgeOldSecretsAsync(IList<string> currentFunctions, ILogger logger)
+        {
+            // no-op - allow stale secrets to remain
+            await Task.Yield();
+        }
+
+        public override Task WriteSnapshotAsync(ScriptSecretsType type, string functionName, ScriptSecrets secrets)
+        {
+            //Runtme is not responsable for encription so this code will never be executed.
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs]: nit: spelling here and below

File: src/WebJobs.Script/Environment/EnvironmentSettingNames.cs [hunk 8]
```diff
@@ -30,6 +30,8 @@ public static class EnvironmentSettingNames
         public const string SkipSslValidation = "SCM_SKIP_SSL_VALIDATION";
         public const string EnvironmentNameKey = "AZURE_FUNCTIONS_ENVIRONMENT";
         public const string EasyAuthEnabled = "WEBSITE_AUTH_ENABLED";
+        public const string AzureWebJobsSecretStorageVaultName = "AzureWebJobsSecretStorageVaultName";
+        public const string AzureWebJobsSecretStorageAuthConnectionString = "AzureWebJobsSecretStorageAuthConnectionString";
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentSettingNames.cs]: I think these setting names should be:

AzureWebJobsSecretStorageKeyVaultName
AzureWebJobsSecretStorageKeyVaultConnectionString

**Comment by mathewc** [comment]
Still some unresolved comments in this PR?

**Comment by alrod** [comment]
> Still some unresolved comments in this PR?

indeed `"Microsoft.KeyVault"` was not changed to `"keyvault"`

**Pull request merged**